src/types/simple.rs

```diff
@@ -272,24 +274,30 @@ impl<'a> TryFrom<&'a SimpleType> for &'a ClassicType {
 
 /// List of types, used for function signatures.
 #[derive(Clone, PartialEq, Eq, Debug, serde::Serialize, serde::Deserialize)]
-#[cfg_attr(feature = "pyo3", pyclass)]
+//#[cfg_attr(feature = "pyo3", pyclass)] // TODO: expose unparameterized versions
 #[non_exhaustive]
 #[serde(transparent)]
-pub struct TypeRow {
+pub struct TypeRow<T: PrimType> {
     /// The datatypes in the row.
-    types: Cow<'static, [SimpleType]>,
+    types: Cow<'static, [T]>,
 }
 
-impl Display for TypeRow {
+/// A row of [SimpleType]s
+pub type SimpleRow = TypeRow<SimpleType>;
+
+/// A row of [ClassicType]s
+pub type ClassicRow = TypeRow<ClassicType>;
+
+impl<T: Display + PrimType> Display for TypeRow<T> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.write_char('[')?;
         display_list(self.types.as_ref(), f)?;
         f.write_char(']')
     }
 }
 
-#[cfg_attr(feature = "pyo3", pymethods)]
-impl TypeRow {
+//#[cfg_attr(feature = "pyo3", pymethods)] // TODO: expose unparameterized versions
```

Comment:
We'll need a wrapper struct for pyo3, so may as well move `len` and `is_empty` to the impl block below.

Reply:
Fair! Done